client/state/imports/reducer.js

```diff
@@ -0,0 +1,290 @@
+				importerState: appStates.READY_FOR_UPLOAD,
+				site: { ID: siteId },
+			} )
+		)( state ),
+	[ IMPORTS_START_IMPORTING ]: ( state, { importerId } ) =>
+		mergeImporterUpdate(
+			importerId,
+			{
+				importerState: appStates.IMPORTING,
+			},
+			state
+		),
+	[ IMPORTS_IMPORT_CANCEL ]: ( state, { importerId } ) => deleteImporter( importerId, state ),
+	[ IMPORTS_IMPORT_RESET ]: ( state, { importerId } ) => deleteImporter( importerId, state ),
+	[ IMPORTS_IMPORT_RECEIVE ]: ( state, { importerStatus } ) => {
+		const importerId = importerStatus.importerId;
+		const updatedState = mergeApiUpdate(
+			{
+				isHydrated: true,
+			},
+			state
+		);
+
+		if ( state.importerLocks[ importerId ] ) {
+			return updatedState;
+		}
+
+		if ( importerStatus.importerState === appStates.DEFUNCT ) {
+			return deleteImporter( importerId, updatedState );
+		}
+
+		return flow( deleteRemovableImporters, createImporter( importerId, importerStatus ) )(
+			updatedState
+		);
+	},
+	[ IMPORTS_UPLOAD_START ]: ( state, { importerId, filename } ) =>
+		mergeImporterUpdate(
+			importerId,
+			{
+				importerState: appStates.UPLOADING,
+				filename,
+			},
+			state
+		),
+	[ IMPORTS_UPLOAD_FAILED ]: ( state, { importerId, error } ) =>
+		mergeImporterUpdate(
+			importerId,
+			{
+				importerState: appStates.UPLOAD_FAILURE,
+				errorData: {
+					type: 'uploadError',
+					description: error,
+				},
+			},
+			state
+		),
+	[ IMPORTS_UPLOAD_SET_PROGRESS ]: ( state, { importerId, uploadLoaded, uploadTotal } ) =>
+		mergeImporterUpdate(
+			importerId,
+			{
+				percentComplete: uploadLoaded / ( uploadTotal + Number.EPSILON ) * 100,
+			},
+			state
+		),
+	[ IMPORTS_UPLOAD_COMPLETED ]: ( state, { importerId, importerStatus } ) =>
+		flow(
+			deleteImporter( importerId ),
+			createImporter( importerStatus.importerId, importerStatus )
+		)( state ),
+	[ IMPORTS_AUTHORS_START_MAPPING ]: ( state, { importerId } ) =>
+		mergeImporterUpdate(
+			importerId,
+			{
+				importerState: appStates.MAP_AUTHORS,
+			},
+			state
+		),
+	[ IMPORTS_AUTHORS_SET_MAPPING ]: ( state, { importerId, sourceAuthor, targetAuthor } ) => {
+		const authors = get( state, [ 'importers', importerId, 'customData', 'sourceAuthors' ] );
+
+		if ( ! authors ) {
+			return state;
+		}
+
+		const mappedAuthors = authors.map( author => {
+			if ( author.id !== sourceAuthor.id ) {
+				return author;
+			}
+
+			return {
+				...author,
+				mappedTo: targetAuthor,
+			};
+		} );
+
+		return mergeImporterUpdate(
+			importerId,
+			{
+				customData: {
+					...state.importers[ importerId ].customData,
+					sourceAuthors: mappedAuthors,
+				},
+			},
+			state
+		);
+	},
+	[ IMPORTS_IMPORT_LOCK ]: ( state, { importerId } ) => ( {
+		...state,
+		importerLocks: {
+			...state.importerLocks,
+			[ importerId ]: true,
+		},
+	} ),
+	[ IMPORTS_IMPORT_UNLOCK ]: ( state, { importerId } ) => ( {
+		...state,
+		importerLocks: {
+			...state.importerLocks,
+			[ importerId ]: false,
+		},
+	} ),
```

Comment:
With `immutable-js`'s built-in helper functions having a single large state object is a bit easier, but here we end up with lots of duplication with the spread operator and one big reducer. We prefer building our normal reducers with composition these days with `combineReducers()` and allow for each sub-reducer to handle its own input. it would be nice to explore this idea now while rewriting all the state code

Reply:
Yeah, I agree the state/reducers could do with a bit of a revamp now that they're in redux.

I thought it might be biting off a bit much to tackle all that in the one PR (I'm a trialist, so fairly new to the codebase).

Reply:
oh that's fine - actually I was hoping to suggest that we break this PR down - it's much much bigger than we usually like to merge, and I'm always advocating for finding ways to refactor things gradually and slowly instead of all-at-once.

that is, maybe we can find small ways to start updating this code in pieces, merge them one at a time, give each change a due chance for review and testing, then move to the next one…

Reply:
It'd be quite hard to migrate the state piece by piece, but the PR could be divided up into one PR for actions, one for reducers, and a third and final PR to hook it all up (and remove the old code). That's kind of how I did my commits anyway.

It would make it more reviewable with the only downside being that the first two PRs would add unused code that couldn't receive any manual tests (but at least it'll be code with unit tests).

Reply:
did you consider breaking apart the state structure and porting over parts of the state at a time?

for example, one PR to create the basic reducer and include in it one data item and the action types relevant to it. that is, create some _temporary_ duplication in the code in order to gradually shift the data structure

```js
this.props.oldFluxyUpdate();
this.props.update() // new redux action dispatcher
```

Reply:
If I'm understanding the proposal, I think it could be complicated by the way the imports state is structured. Many of the action creators trigger multiple actions and the reducers read from and update multiple parts of the importer state, they're not isolated in a way that would make it easy to have some implemented in flux and some in redux.

I think the options to proceed along those lines would be either:
a) refactor the state so that it can be migrated piece by piece (and abandoning this PR)
b) to have some way to make flux read from particular parts of redux's state tree (sounds scary, but maybe you've used similar approaches in the past?) and breaking this PR up.

Then again, I might be completely misunderstanding.

Reply:
1500 lines isn't unheard of, but it's not really about the size so much as the surface area for bugs and the ability to review and find them.

option (a) sounds legitimate; this PR could be a good guide for splitting up the work. ideally it would be great if we could propose a PR that makes one single change and needs one specific thing to test, verses one that changes an entire subsystem and whose testing instructions are "make sure imports still work"

in this case maybe the risk is lower because of the exist code? maybe not? I was hoping that we could pull out something, for example the author mapping, into a separate reducer, etc… and move gradually.

anyway, thanks for evaluating this. I'd still propose the challenge and suggest you explore it a little bit on how possible it is to break apart, but if you end up thinking it will take three times as long (for example) then let pragmatism be your guide. just remember that usually we underestimate how long single large atomic changes can take :wink:

Reply:
> to have some way to make flux read from particular parts of redux's state tree

we have a module called the [redux-bridge](https://github.com/Automattic/wp-calypso/blob/master/client/lib/redux-bridge/index.js)

Reply:
Thanks for the pointer - I'm just looking into this to see if there's a way I can break this up into incremental PRs.

Reply:
@talldan I do similar migrations as a series of smaller PRs that go like this:
1. Add the Redux store with actions, reducers and selectors. It's not used anywhere yet. It's tested only by unit tests. It's usually a large diff, but it's not dangerous if some bug is not catched in this phase, as the code is not yet used.
2. Use the new Redux store instead of Flux in the app code. This phase needs to be reviewed and tested most carefully, and it helps that the diff is usually rather small.
3. Remove the old Flux code that's no longer used. Large diff, but trivial to review.

The most important benefit is that the reviewer doesn't need to scroll through large diffs like the current PR. The risky part is concentrated in a small diff in part 2.

Reply:
I'd like to add that it's nice to break step (1) into small pieces as well, and if possible, run several flows through 1/2/3 :smile: